ports/stm/supervisor/port.c

```diff
@@ -106,3 +140,109 @@ void HardFault_Handler(void) {
+        uint8_t day = (uint8_t)(date & (RTC_DR_DT | RTC_DR_DU));
+        // Add 2000 since the year is only the last two digits.
+        year = 2000 + (uint32_t)RTC_Bcd2ToByte(year);
+        month = (uint8_t)RTC_Bcd2ToByte(month);
+        day = (uint8_t)RTC_Bcd2ToByte(day);
+        seconds_to_date = timeutils_seconds_since_2000(year, month, day, 0, 0, 0);
+        cached_date = date;
+    }
+    uint32_t hours_minutes = time & (RTC_TR_HT | RTC_TR_HU | RTC_TR_MNT | RTC_TR_MNU);
+    if (hours_minutes != cached_hours_minutes) {
+        uint8_t hours = (uint8_t)((time & (RTC_TR_HT | RTC_TR_HU)) >> 16U);
+        uint8_t minutes = (uint8_t)((time & (RTC_TR_MNT | RTC_TR_MNU)) >> 8U);
+        hours = (uint8_t)RTC_Bcd2ToByte(hours);
+        minutes = (uint8_t)RTC_Bcd2ToByte(minutes);
+        seconds_to_minute = 60 * (60 * hours + minutes);
+    }
+    uint8_t seconds = (uint8_t)(time & (RTC_TR_ST | RTC_TR_SU));
+    seconds = (uint8_t)RTC_Bcd2ToByte(seconds);
+    if (subticks != NULL) {
+        *subticks = subseconds % 32;
+    }
+
+    return ((uint64_t) 1024) * (seconds_to_date + seconds_to_minute + seconds) + subseconds / 32;
+}
+
+void RTC_WKUP_IRQHandler(void) {
+    supervisor_tick();
+    __HAL_RTC_WAKEUPTIMER_CLEAR_FLAG(&_hrtc, RTC_FLAG_WUTF);
+    __HAL_RTC_WAKEUPTIMER_EXTI_CLEAR_FLAG();
+}
+volatile bool alarmed_already = false;
+void RTC_Alarm_IRQHandler(void) {
+    __HAL_RTC_ALARM_CLEAR_FLAG(&_hrtc, RTC_FLAG_ALRAF);
+    __HAL_RTC_ALARM_EXTI_CLEAR_FLAG();
+    HAL_RTC_DeactivateAlarm(&_hrtc, RTC_ALARM_A);
+    alarmed_already = true;
+}
+
+// Enable 1/1024 second tick.
+void port_enable_tick(void) {
+    HAL_RTCEx_SetWakeUpTimer_IT(&_hrtc, RTC_CLOCK_FREQUENCY / 1024 / 2, RTC_WAKEUPCLOCK_RTCCLK_DIV2);
+    HAL_NVIC_SetPriority(RTC_WKUP_IRQn, 1, 0U);
+    HAL_NVIC_EnableIRQ(RTC_WKUP_IRQn);
+}
+extern volatile uint32_t autoreload_delay_ms;
+
+// Disable 1/1024 second tick.
+void port_disable_tick(void) {
+    HAL_NVIC_DisableIRQ(RTC_WKUP_IRQn);
+    HAL_RTCEx_DeactivateWakeUpTimer(&_hrtc);
+}
+
+void port_interrupt_after_ticks(uint32_t ticks) {
+    uint64_t raw_ticks = port_get_raw_ticks(NULL) + ticks;
+
+    RTC_AlarmTypeDef alarm;
+    if (ticks > 1024) {
+        timeutils_struct_time_t tm;
+        timeutils_seconds_since_2000_to_struct_time(raw_ticks / 1024, &tm);
+        alarm.AlarmTime.Hours = tm.tm_hour;
+        alarm.AlarmTime.Minutes = tm.tm_min;
+        alarm.AlarmTime.Seconds = tm.tm_sec;
+        alarm.AlarmDateWeekDay = tm.tm_mday;
+        alarm.AlarmMask = RTC_ALARMMASK_ALL;
+    } else {
+        alarm.AlarmMask = RTC_ALARMMASK_NONE;
+    }
+
+    alarm.AlarmTime.SubSeconds = RTC_CLOCK_FREQUENCY -
+                                 ((raw_ticks % (1024)) * 32);
+    alarm.AlarmTime.DayLightSaving = RTC_DAYLIGHTSAVING_NONE;
+    alarm.AlarmTime.StoreOperation = RTC_STOREOPERATION_SET;
+    alarm.AlarmSubSecondMask = RTC_ALARMSUBSECONDMASK_NONE;
+    alarm.AlarmDateWeekDaySel = RTC_ALARMDATEWEEKDAYSEL_DATE;
```

Comment:
should this be set when ticks > 1000 only?

Reply:
Ah, I understand now.  I had a wrong guess about what the flag meant, and didn't verify it.